lib/Sema/ConstraintSystem.cpp

```diff
@@ -1175,7 +1175,8 @@ ConstraintSystem::getTypeOfMemberReference(
          func->hasArchetypeSelf())) {
       openedType = openedType->replaceCovariantResultType(
                      baseObjTy,
-                     func->getNumParameterLists());
+                     func->getNumParameterLists(),
+                     !baseObjTy->getAnyOptionalObjectType());
```

Comment:
I think it's valid to remove `assert` on optionality of the `newResultType` since there are valid cases for that like what we both mentioned, WDYT, @DougGregor?